torch/csrc/utils/tensor_new.cpp

```diff
@@ -651,18 +644,55 @@ Tensor sparse_csr_tensor_ctor(c10::DispatchKey dispatch_key, at::ScalarType scal
                                            r.pyobject(VALUES_ARG), /*copy_variables=*/false, /*copy_numpy=*/true,
                                            /*type_inference=*/type_inference);
     Tensor crow_indices = internal_new_from_data(values.options(),
-      crow_indices_scalar_type, r.deviceOptional(DEVICE_TYPE_ARG),
+      kInt, r.deviceOptional(DEVICE_TYPE_ARG),
       r.pyobject(CROW_INDICES_ARG), /*copy_variables=*/false, /*copy_numpy=*/true,
-      /*type_inference=*/false);
-    Tensor col_indices = internal_new_from_data(values.options(), col_indices_scalar_type, r.deviceOptional(DEVICE_TYPE_ARG),
+      /*type_inference=*/true);
+    Tensor col_indices = internal_new_from_data(values.options(), kInt, r.deviceOptional(DEVICE_TYPE_ARG),
       r.pyobject(COL_INDICES_ARG), /*copy_variables=*/false, /*copy_numpy=*/true,
-      /*type_inference=*/false);
+      /*type_inference=*/true);
     return at::sparse_csr_tensor(crow_indices, col_indices, values,
                                  values.options().layout(at::kSparseCsr)).set_requires_grad(r.toBool(REQ_GRAD_ARG));
   }
   throw std::runtime_error("sparse_csr_tensor(): invalid arguments");
 }
 
+Tensor _sparse_csr_tensor_unsafe_ctor(c10::DispatchKey dispatch_key, at::ScalarType scalar_type, PyObject* args, PyObject* kwargs) {
+  TORCH_INTERNAL_ASSERT(!isSparseCsr(dispatchKeyToBackend(dispatch_key)));
```

Comment:
Probably want to check `!isSparse()` here too, right? (And similarly add `!isSparseCsr()` to the COO version)

Reply:
Sure, just to be sure that everything is ok these checks are added too.

Reply:
> Probably want to check `!isSparse()` here too, right? (And similarly add `!isSparseCsr()` to the COO version)

Why? `Sparse` corresponds to `COO` layout and `SparseCsr` corresponds to `CSR` layout. The different layouts should not infer with each other, IIUC.